authz/sdk_end2end_test.go

```diff
@@ -305,3 +379,168 @@ func TestSDKEnd2End(t *testing.T) {
 		})
 	}
 }
+
+func (s) TestSDKFileWatcher_ValidPolicyRefresh(t *testing.T) {
+	valid1 := sdkTests["DeniesRpcMatchInDenyAndAllow"]
+	file := createTmpPolicyFile(t, "valid_policy_refresh", []byte(valid1.authzPolicy))
+	i, _ := authz.NewFileWatcher(file, 1*time.Second)
+	defer i.Close()
+
+	// Start a gRPC server with SDK unary server interceptor.
+	s := grpc.NewServer(
+		grpc.ChainUnaryInterceptor(i.UnaryInterceptor))
+	defer s.Stop()
+	pb.RegisterTestServiceServer(s, &testServer{})
+
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("error listening: %v", err)
+	}
+	defer lis.Close()
+	go s.Serve(lis)
+
+	// Establish a connection to the server.
+	clientConn, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("grpc.Dial(%v) failed: %v", lis.Addr().String(), err)
+	}
+	defer clientConn.Close()
+	client := pb.NewTestServiceClient(clientConn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	// Verifying authorization decision.
+	_, err = client.UnaryCall(ctx, &pb.SimpleRequest{})
+	if got := status.Convert(err); got.Code() != valid1.wantStatusCode || got.Message() != valid1.wantErr {
+		t.Fatalf("error want:{%v %v} got:{%v %v}", valid1.wantStatusCode, valid1.wantErr, got.Code(), got.Message())
+	}
+
+	// Rewrite the file with a different valid authorization policy.
+	valid2 := sdkTests["AllowsRpcEmptyDenyMatchInAllow"]
+	if err := ioutil.WriteFile(file, []byte(valid2.authzPolicy), os.ModePerm); err != nil {
+		t.Fatalf("ioutil.WriteFile(%q) failed: %v", file, err)
+	}
+	// Wait 2 seconds for background go routine to read the updated files.
+	time.Sleep(2 * time.Second)
```

Comment:
Sleeps in tests for this purpose are race-prone (if too short) OR make tests unnecessarily slow (if too long).  Unfortunately it's impossible to find the exact right amount of time to wait.

Normally what we'd do is have a blocking way to detect the update happened.  Since this is an API-level test, and such an API is undesirable, instead what you should do is perform `UnaryCall` in a loop.  _Accept_ the old behavior, but break out of the loop when the new behavior is detected (and fail if some other behavior is observed, if applicable).  You can give that loop ~10 seconds before failing due to a timeout of not seeing the desired behavior.  To avoid a tight busy-loop burning CPU cycles needlessly, sleep for something like 10ms at the end of the loop.

Reply:
Updated the tests to loop for desired behavior. Also reduced the refresh duration from 1 second to 10 millisecond. PTAL.